File: backend/app/task_engine.py
import hashlib
from dataclasses import dataclass
from datetime import date, datetime, time, timedelta, timezone

from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.config import Settings
from app.models import (
    Activity,
    ActivityStatus,
    CropPeriod,
    CropPeriodStatus,
    Farm,
    Task,
    TaskConfidence,
    TaskPriority,
    TaskSource,
    TaskStatus,
    WeatherSnapshot,
    utcnow,
)
from app.risk_rules import RiskSeverity, WeatherRisk, evaluate_weather_risks
from app.weather import WeatherProviderError, deserialize_weather_points


@dataclass(frozen=True)
class TaskSpec:
    title: str
    description: str
    reason: str
    priority: TaskPriority
    source: TaskSource
    confidence: TaskConfidence
    due_date: date
    crop_period_id: object | None = None
    dedupe_discriminator: str = ""


def ensure_daily_tasks(
    db: Session,
    farm: Farm,
    target_date: date,
    settings: Settings,
) -> list[Task]:
    if target_date != date.today() or farm.archived_at is not None:
        return []

    crop_period = db.scalar(
        select(CropPeriod).where(
            CropPeriod.farm_id == farm.id,
            CropPeriod.status == CropPeriodStatus.ACTIVE,
        )
    )
    specs: list[TaskSpec] = []
    if crop_period is not None:
        growing_day = max((target_date - crop_period.planted_at).days + 1, 1)
        specs.append(
            TaskSpec(
                title="Ürün gelişimini sahada kontrol edin",
                description=(
                    "Bitki gelişimini, toprak nemini ve olağan dışı belirtileri "
                    "kontrol edip gözlemlerinizi tarla günlüğüne kaydedin."
                ),
                reason=(
                    f"{crop_period.crop_type.value} üretim döneminin "
                    f"{growing_day}. günü için düzenli saha kontrolü."
                ),
                priority=TaskPriority.MEDIUM,
                source=TaskSource.CROP_CALENDAR,
                confidence=TaskConfidence.MEDIUM,
                due_date=target_date,
                crop_period_id=crop_period.id,
                dedupe_discriminator="daily-field-check",
            )
        )

    since = datetime.combine(
        target_date - timedelta(days=7),
        time.min,
        tzinfo=timezone.utc,
    )
    recent_activity = db.scalar(
        select(Activity.id)
        .where(
            Activity.farm_id == farm.id,
            Activity.status == ActivityStatus.CONFIRMED,
            Activity.archived_at.is_(None),
            Activity.occurred_at >= since,
        )
        .limit(1)
    )
    if recent_activity is None:
        specs.append(
            TaskSpec(
                title="Tarla günlüğünü güncelleyin",
                description=(
                    "Son sulama, gübreleme, ilaçlama veya saha kontrolü gibi "
                    "işlemleri tarla günlüğüne ekleyin."
                ),
                reason="Son yedi gün içinde doğrulanmış faaliyet kaydı bulunamadı.",
                priority=TaskPriority.LOW,
                source=TaskSource.SYSTEM,
                confidence=TaskConfidence.HIGH,
                due_date=target_date,
                crop_period_id=crop_period.id if crop_period else None,
                dedupe_discriminator="weekly-activity-reminder",
            )
        )

    specs.extend(
        _fresh_weather_task_specs(
            db,
            farm,
            crop_period,
            target_date,
            settings,
        )
    )

    created: list[Task] = []
    for spec in specs:
        dedupe_key = task_dedupe_key(
            source=spec.source,
            title=spec.title,
            description=spec.description,
            reason=spec.reason,
            crop_period_id=spec.crop_period_id,
            discriminator=spec.dedupe_discriminator,
        )
        exists = db.scalar(
            select(Task.id).where(
                Task.farm_id == farm.id,
                Task.due_date == target_date,
                Task.dedupe_key == dedupe_key,
            )
        )
        if exists is not None:
            continue
        task = Task(
            farm_id=farm.id,
            crop_period_id=spec.crop_period_id,
            title=spec.title,
            description=spec.description,
            reason=spec.reason,
            priority=spec.priority,
            status=TaskStatus.NEW,
            source=spec.source,
            confidence=spec.confidence,
            due_date=spec.due_date,
            dedupe_key=dedupe_key,
        )
        db.add(task)
        created.append(task)

    if created:
        try:
            db.commit()
        except IntegrityError:
            # Another request may have generated the same daily set first.
            # The unique key remains the source of truth; the caller re-queries.
            db.rollback()
            return []
        for task in created:
            db.refresh(task)
    return created


def task_dedupe_key(
    *,
    source: TaskSource,
    title: str,
    description: str,
    reason: str,
    crop_period_id: object | None,
    discriminator: str = "",
) -> str:
    normalized = "|".join(
        (
            source.value,
            " ".join(title.casefold().split()),
            " ".join(description.casefold().split()),
            " ".join(reason.casefold().split()),
            str(crop_period_id or ""),
            discriminator,
        )
    )
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()


def mark_overdue_tasks(db: Session, farm_id: object, before: date) -> None:
    tasks = db.scalars(
        select(Task).where(
            Task.farm_id == farm_id,
            Task.due_date < before,
            Task.status.in_(
                (TaskStatus.NEW, TaskStatus.VIEWED, TaskStatus.PLANNED)
            ),
        )
    ).all()
    if not tasks:
        return
    for task in tasks:
        task.status = TaskStatus.OVERDUE
    db.commit()


def task_sort_key(task: Task) -> tuple[int, int, datetime]:
    priority_rank = {
        TaskPriority.CRITICAL: 0,
        TaskPriority.HIGH: 1,
        TaskPriority.MEDIUM: 2,
        TaskPriority.LOW: 3,
    }
    source_rank = {
        TaskSource.EXPERT: 0,
        TaskSource.WEATHER: 1,
        TaskSource.CROP_CALENDAR: 2,
        TaskSource.SYSTEM: 3,
    }
    return (
        priority_rank[task.priority],
        source_rank[task.source],
        _as_utc(task.created_at),
    )


def _fresh_weather_task_specs(
    db: Session,
    farm: Farm,
    crop_period: CropPeriod | None,
    target_date: date,
    settings: Settings,
) -> list[TaskSpec]:
    snapshot = db.scalar(
        select(WeatherSnapshot)
        .where(WeatherSnapshot.farm_id == farm.id)
        .order_by(WeatherSnapshot.fetched_at.desc())
        .limit(1)
    )
    if snapshot is None:
        return []
    fetched_at = _as_utc(snapshot.fetched_at)
    if fetched_at < utcnow() - timedelta(hours=settings.weather_stale_after_hours):
        return []
    try:
        risks = evaluate_weather_risks(
            deserialize_weather_points(snapshot.payload),
            now=utcnow(),
        )
    except WeatherProviderError:
        return []
    return [
        _weather_task_spec(risk, crop_period, target_date, snapshot)
        for risk in risks
    ]


def _weather_task_spec(
    risk: WeatherRisk,
    crop_period: CropPeriod | None,
    target_date: date,
    snapshot: WeatherSnapshot,
) -> TaskSpec:
    title_by_risk = {
        "FROST": "Don riskine karşı tarlanızı kontrol edin",
        "STRONG_WIND": "Kuvvetli rüzgâr riskini değerlendirin",
        "HEAVY_RAIN": "Yoğun yağış riskini değerlendirin",
    }
    return TaskSpec(
        title=title_by_risk[risk.risk_type.value],
        description=risk.suggested_action,
        reason=(
            f"{risk.message} Hava verisi {snapshot.provider} tarafından "
            f"{_as_utc(snapshot.fetched_at).isoformat()} tarihinde güncellendi; "
            "tahminler değişebilir."
        ),
        priority=(
            TaskPriority.CRITICAL
            if risk.severity == RiskSeverity.CRITICAL
            else TaskPriority.HIGH
        ),
        source=TaskSource.WEATHER,
        confidence=TaskConfidence.MEDIUM,
        due_date=target_date,
        crop_period_id=crop_period.id if crop_period else None,
        dedupe_discriminator=risk.risk_type.value,
    )


def _as_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)

File: backend/src/TarlaAsistani.Application/Features/Tasks/Services/TaskEngine.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using TarlaAsistani.Application.Common.Interfaces;
using TarlaAsistani.Application.Features.Weather.Services;
using TarlaAsistani.Domain.Entities;
using TarlaAsistani.Domain.Enums;
using TaskStatus = TarlaAsistani.Domain.Enums.TaskStatus;

namespace TarlaAsistani.Application.Features.Tasks.Services;

public record TaskSpec(
    string Title,
    string Description,
    string Reason,
    TaskPriority Priority,
    TaskSource Source,
    TaskConfidence Confidence,
    DateOnly DueDate,
    Guid? CropPeriodId = null,
    string DedupeDiscriminator = ""
);

public static class TaskEngine
{
    public static async Task EnsureDailyTasksAsync(
        IApplicationDbContext db,
        Farm farm,
        DateOnly targetDate,
        CancellationToken cancellationToken = default)
    {
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        if (targetDate != today || farm.ArchivedAt != null)
        {
            return;
        }

        var cropPeriod = await db.CropPeriods
            .FirstOrDefaultAsync(cp => cp.FarmId == farm.Id && cp.Status == CropPeriodStatus.Active, cancellationToken);

        var specs = new List<TaskSpec>();

        // 1. Crop calendar growth check
        if (cropPeriod != null)
        {
            var growingDay = Math.Max(targetDate.DayNumber - cropPeriod.PlantedAt.DayNumber + 1, 1);
            var cropDisplayName = !string.IsNullOrWhiteSpace(cropPeriod.CropName)
                ? cropPeriod.CropName
                : cropPeriod.CropType?.ToString();
            specs.Add(new TaskSpec(
                Title: "Ürün gelişimini sahada kontrol edin",
                Description: "Bitki gelişimini, toprak nemini ve olağan dışı belirtileri kontrol edip gözlemlerinizi tarla günlüğüne kaydedin.",
                Reason: $"{cropDisplayName} üretim döneminin {growingDay}. günü için düzenli saha kontrolü.",
                Priority: TaskPriority.Medium,
                Source: TaskSource.CropCalendar,
                Confidence: TaskConfidence.Medium,
                DueDate: targetDate,
                CropPeriodId: cropPeriod.Id,
                DedupeDiscriminator: "daily-field-check"
            ));
        }

        // 2. Weekly activity reminder
        var sevenDaysAgo = DateTime.UtcNow.AddDays(-7);
        var recentActivityExists = await db.Activities
            .AnyAsync(a => a.FarmId == farm.Id &&
                           a.Status == ActivityStatus.Confirmed &&
                           a.ArchivedAtUtc == null &&
                           a.OccurredAtUtc >= sevenDaysAgo, cancellationToken);

        if (!recentActivityExists)
        {
            specs.Add(new TaskSpec(
                Title: "Tarla günlüğünü güncelleyin",
                Description: "Son sulama, gübreleme, ilaçlama veya saha kontrolü gibi işlemleri tarla günlüğüne ekleyin.",
                Reason: "Son yedi gün içinde doğrulanmış faaliyet kaydı bulunamadı.",
                Priority: TaskPriority.Low,
                Source: TaskSource.System,
                Confidence: TaskConfidence.High,
                DueDate: targetDate,
                CropPeriodId: cropPeriod?.Id,
                DedupeDiscriminator: "weekly-activity-reminder"
            ));
        }

        // 3. Weather risk tasks from fresh weather snapshot
        var latestSnapshot = await db.WeatherSnapshots
            .Where(s => s.FarmId == farm.Id)
            .OrderByDescending(s => s.FetchedAtUtc)
            .FirstOrDefaultAsync(cancellationToken);

        if (latestSnapshot != null && latestSnapshot.FetchedAtUtc >= DateTime.UtcNow.AddHours(-4))
        {
            try
            {
                var points = JsonSerializer.Deserialize<List<WeatherPoint>>(latestSnapshot.Payload);
                if (points != null && points.Count > 0)
                {
                    var risks = WeatherRiskEvaluator.Evaluate(points, DateTime.UtcNow);
                    foreach (var risk in risks)
                    {
                        var title = risk.RiskType switch
                        {
                            "FROST" => "Don riskine karşı tarlanızı kontrol edin",
                            "STRONG_WIND" => "Kuvvetli rüzgâr riskini değerlendirin",
                            "HEAVY_RAIN" => "Yoğun yağış riskini değerlendirin",
                            _ => "Hava koşullarını değerlendirin"
                        };

                        specs.Add(new TaskSpec(
                            Title: title,
                            Description: risk.SuggestedAction,
                            Reason: $"{risk.Message} Hava verisi {latestSnapshot.Provider} tarafından güncellendi.",
                            Priority: risk.Severity == "CRITICAL" ? TaskPriority.Critical : TaskPriority.High,
                            Source: TaskSource.Weather,
                            Confidence: TaskConfidence.Medium,
                            DueDate: targetDate,
                            CropPeriodId: cropPeriod?.Id,
                            DedupeDiscriminator: risk.RiskType
                        ));
                    }
                }
            }
            catch
            {
                // Ignore payload parsing errors in background engine
            }
        }

        // 4. Create tasks with deduplication
        var distinctSpecs = specs
            .GroupBy(s => CalculateDedupeKey(s))
            .Select(g => g.First())
            .ToList();

        var activeOrOverdueStatuses = new[]
        {
            TaskStatus.New,
            TaskStatus.Viewed,
            TaskStatus.Planned,
            TaskStatus.Overdue
        };

        var now = DateTime.UtcNow;
        foreach (var spec in distinctSpecs)
        {
            var dedupeKey = CalculateDedupeKey(spec);
            var legacyDedupeKey = CalculateLegacyDedupeKey(spec);

            // Prevent duplicate creation:
            // 1. If task with same dedupeKey (new or legacy) already exists for targetDate (any status)
            // 2. OR if an active/overdue task with same dedupeKey (new or legacy) already exists for this farm
            var alreadyExists = await db.FarmTasks
                .AnyAsync(t => t.FarmId == farm.Id &&
                               (t.DedupeKey == dedupeKey || t.DedupeKey == legacyDedupeKey) &&
                               (t.DueDate == targetDate || activeOrOverdueStatuses.Contains(t.Status)),
                          cancellationToken);

            if (!alreadyExists)
            {
                db.FarmTasks.Add(new FarmTask
                {
                    Id = Guid.NewGuid(),
                    FarmId = farm.Id,
                    Title = spec.Title,
                    Description = spec.Description,
                    Reason = spec.Reason,
                    Priority = spec.Priority,
                    Source = spec.Source,
                    Confidence = spec.Confidence,
                    Status = TaskStatus.New,
                    DueDate = spec.DueDate,
                    CropPeriodId = spec.CropPeriodId,
                    DedupeKey = dedupeKey,
                    CreatedAtUtc = now,
                    UpdatedAtUtc = now
                });
            }
        }

        await db.SaveChangesAsync(cancellationToken);
    }

    public static string CalculateDedupeKey(TaskSpec spec)
    {
        var discriminator = string.IsNullOrWhiteSpace(spec.DedupeDiscriminator)
            ? spec.Title.Trim().ToLowerInvariant()
            : spec.DedupeDiscriminator.Trim().ToLowerInvariant();
        var raw = $"{spec.Source}|{discriminator}|{spec.CropPeriodId}";
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(raw))).ToLowerInvariant();
    }

    public static string CalculateLegacyDedupeKey(TaskSpec spec)
    {
        var raw = $"{spec.Source}|{spec.Title.ToLowerInvariant()}|{spec.Description.ToLowerInvariant()}|{spec.Reason.ToLowerInvariant()}|{spec.CropPeriodId}|{spec.DedupeDiscriminator}";
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(raw))).ToLowerInvariant();
    }
}
